src/sp/it/pl/audio/playlist/PlaylistItem.kt

```diff
@@ -121,7 +121,7 @@ class PlaylistItem: Item {
 
         if (isFileBased()) {
             throwIfFxThread()
-            getFile().readAudioFile().orNull()?.let { f ->
+            getFile()!!.readAudioFile().orNull()?.let { f ->
```

Comment:
You are right to point out the problem and I was hesitant to write it that way, but decided in favor because it is more readable than just checking null - that hides the logic of the decision from the reader and puts the very existence of isFileBased() method to question. Consuming the null from getFile() is fine, but not using it to perform the isFileBased() check - there is a method for that with a clear name and all! This is also premature optimization. The call is non blocking and in order of ns. Consumer of the API also shouldn't know the call is repeated and it may not be in the future.

Once we update Kotlin to 1.3, isFileBased() will receive contract that will specify that if it returns true, getFile() returns non null File. This will remove usages of !! after getFile() after isFileBased() is called.
Check https://proandroiddev.com/kotlin-contracts-make-great-deals-with-the-compiler-f524e57f11c

I believe that once the contract is there, this solution becomes very elegant. What developer wants to do here is to handle filebased and non file based cases differently -split the logic and this solution does exactly that, explicitly.

As for the getFileDirect() method - im absolutely against. It complicates API and involves consumer with implementation details.

Reply:
Kotlin 1.3 is here! Let's go :)

Reply:
Indeed, but lets leave that up to new PR which will be for updating kotlin only. There is more things that will be involved with that. And I'll work on it during the weekend.